tests/rptest/tests/end_to_end.py

```diff
@@ -257,39 +271,97 @@ def run_consumer_validation(self,
             raise
 
-    def validate(self, enable_idempotence):
+    def validate_compacted(self):
+
+        consumer_state = {}
+
+        acked_producer_state = {}
+        not_acked_producer_state = defaultdict(set)
+
+        for k, v in self.producer.acked:
+            acked_producer_state[k] = v
+
+        # some of the not acked records may have been received and persisted,
+        # we need to store all of them and check if consumer result is one of them
+        for k, v in self.producer.not_acked:
+            not_acked_producer_state[k].add(v)
+
+        for k, v in self.records_consumed:
+            consumer_state[k] = v
+
+        msg = ""
+        success = True
+        errors = []
+
+        for consumed_key, consumed_value in consumer_state.items():
+            # invalid key consumed
+            if consumed_key not in acked_producer_state and consumed_key not in not_acked_producer_state:
+                return False, f"key {consumed_key} was consumed but it is missing in produced state"
+
+            # success case, simply continue
+            if acked_producer_state[consumed_key] == consumed_value:
+                continue
+
+            # we must check not acked state as it might have been caused
+            # by request timeout and a message might still have been consumed by consumer
+            self.logger.debug(
+                f"Checking not acked produced messages for key: {k}, previous acked value: {acked_producer_state[consumed_key]}, consumed value: {v}"
+            )
+            # consumed value is one of the not acked produced values
+            if consumed_key in not_acked_producer_state and consumed_value in not_acked_producer_state[
+                    consumed_key]:
+                continue
+
+            # consumed value is not equal to last acked produced value and any of not acked value, error out
+            success = False
+            errors.append((consumed_key, consumed_value,
+                           acked_producer_state.get(consumed_key, None),
+                           not_acked_producer_state.get(consumed_key, None)))
+
+        if not success:
+            msg += "Invalid value detected for consumed compacted topic records. errors: ["
+            for key, consumed_value, produced_acked, producer_not_acked in errors:
+                msg += f"key: {k} consumed value: {consumed_value}, produced values: (acked: {produced_acked}, not_acked: {producer_not_acked}) "
+            msg += "]"
+
+        return success, msg
+
+    def validate(self, enable_idempotence, enable_compaction):
         self.logger.info("Number of acked records: %d" %
                          len(self.producer.acked))
         self.logger.info("Number of consumed records: %d" %
                          len(self.records_consumed))
 
         success = True
         msg = ""
+        if enable_compaction:
+            success, msg = self.validate_compacted()
+        else:
```

Comment:
nit: just to be sure.. we can have either enable_compaction or enable_idempotence but not both because they can conflict with each other, right? We should probably assert that in run_validation() so someone don't set both to true (feel free to skip if you want to avoid another CI round).

Reply:
Will add this in a follow up PR.